stdlib/public/core/PrintForDebugger.swift

```diff
@@ -0,0 +1,294 @@
+    if let sc = mirror.superclassMirror {
+      return ivarCount(mirror: sc) > 0
+    } else {
+      return true
+    }
+  }
+
+  static func printForDebuggerImpl<StreamType: OutputStream>(
+    value: Any?,
+    mirror: Mirror,
+    name: String?,
+    indent: Int,
+    maxDepth: Int,
+    isRoot: Bool,
+    parentCollectionStatus: CollectionStatus,
+    refsAlreadySeen: inout Set<ObjectIdentifier>,
+    maxItemCounter: inout Int,
+    targetStream: inout StreamType) {
+    if maxItemCounter <= 0 {
+      return
+    }
+
+    if !shouldExpand(mirror: mirror,
+                     collectionStatus: parentCollectionStatus,
+                     isRoot: isRoot) {
+      return
+    }
+
+    maxItemCounter -= 1
+  
+    for _ in 0..<indent {
+      print(" ", terminator: "", to: &targetStream)
+    }
+
+    // do not expand classes with no custom Mirror
+    // yes, a type can lie and say it's a class when it's not since we only
+    // check the displayStyle - but then the type would have a custom Mirror
+    // anyway, so there's that...
+    var willExpand = true
+    if let ds = mirror.displayStyle {
+      if ds == .`class` {
+        if let x = value {
+          if !(x is CustomReflectable) {
+            willExpand = false
+          }
+        }
+      }
+    }
+
+    let count = Int(mirror.children.count)
+    let bullet = isRoot && (count == 0 || willExpand == false) ? ""
+      : count == 0    ? "- "
+      : maxDepth <= 0 ? "▹ " : "▿ "
+    print("\(bullet)", terminator: "", to: &targetStream)
+  
+    let collectionStatus = parentCollectionStatus.getChildStatus(child: mirror)
+  
+    if let nam = name {
+      print("\(nam) : ", terminator: "", to: &targetStream)
+    }
+
+    if let str = asStringRepresentation(value: value, mirror: mirror, count: count) {
+      print("\(str)", terminator: "", to: &targetStream)
+    }
+  
+    if (maxDepth <= 0) || !willExpand {
+      print("", to: &targetStream)
+      return
+    }
+
+    if let x = value {
+      if let valueIdentifier = asObjectIdentifier(value: x) {
+        if refsAlreadySeen.contains(valueIdentifier) {
+          print(" { ... }", to: &targetStream)
+          return
+        } else {
+          refsAlreadySeen.insert(valueIdentifier)
+        }
+      }
+    }
+
+    print("", to: &targetStream)
+  
+    var printedElements = 0
+  
+    if let sc = mirror.superclassMirror {
+      printForDebuggerImpl(value: nil,
+                           mirror: sc,
+                           name: "super",
+                           indent: indent + 2,
+                           maxDepth: maxDepth - 1,
+                           isRoot: false,
+                           parentCollectionStatus: .NotACollection,
+                           refsAlreadySeen: &refsAlreadySeen,
+                           maxItemCounter: &maxItemCounter,
+                           targetStream: &targetStream)
+    }
+  
+    for (optionalName,child) in mirror.children {
```

Comment:
I don't necessarily agree that the standard library and the debugger have matching preferred behaviors; if I `dump()` something, I actually expect to get the full printout of the object, not a truncated version because the object happens to be a class with no custom Mirror.

OTOH, that is the expectation (from Objective-C precedent) for the `po` command.

I'd very strongly like to keep things such that there is a debugger-specific entry point for getting printouts of object. As is, this is not public API so users don't run the risk of accidentally using this instead of proper public entry points via code completion.